tuning/utils/data_utils.py

```diff
@@ -0,0 +1,36 @@
+# Standard
+import re
+
+
+def apply_custom_formatting_template(dataset, template, eos_token=""):
+    """Function to format datasets with Alpaca style / other templates.
+    Args:
+        dataset: the HF Dataset element loaded from a JSON or DatasetDict object.
+        template: Template to format data with. Features of Dataset
+            should be referred to by {{key}}
+        eos_token: string EOS token to be appended while formatting data to a single sequence.
+            Defaults to empty
+    Returns:
+        Formatted HF Dataset, dataset_field name that contains formatted data.
+    """
+
+    formatted_dataset_field = "formatted_data_field"
+    template += eos_token
+
+    def formatter(element):
+        def replace_text(match_obj):
+            captured_groups = match_obj.groups()
+            if len(captured_groups) != 1:
+                raise ValueError(
+                    "Unexpectedly captured multiple groups in template formatting"
+                )
+
+            index_object = captured_groups[0]
+            if index_object not in element:
+                raise KeyError("Requested template string is not a valid key in dict")
+
+            return element[index_object]
+
+        return {formatted_dataset_field: re.sub(r"{{(.+)}}", replace_text, template)}
```

Comment:
Just a thought, matching on `(.+)` is also going to match `{` / `}`, which could cause some slightly weird scenarios if people format templates badly - should we be more restrictive on the templates characters? E.g., `[ 0-9a-zA-Z_-]` or something?

I don't have a strong opinion, but might be a good idea to discuss before merging rather than after

Reply:
the regex here is used to find fields of JSON , after field is found the text is put into the template.if people have / in JSON field - "{{input/text}}" , I think its ok to allow it?

Reply:
oh that wont work, hmm ya I can restrict and document. doing that now

Reply:
tested various edge cases and it worked ok

Reply:
I pushed change anyway to be safe and updated README